cvat/apps/consensus/new_intersect_merge.py

```diff
@@ -0,0 +1,1107 @@
+    distance_index = attrib(converter=dict, default={})
+
+    def _instance_bbox(
+        self, instance_anns: Sequence[dm.Annotation]
+    ) -> Tuple[float, float, float, float]:
+        return dm.ops.max_bbox(
+            a.get_bbox() if isinstance(a, dm.Skeleton) else a
+            for a in instance_anns
+            if hasattr(a, "get_bbox") and not a.attributes.get("outside", False)
+        )
+
+    def distance(a, b):
+        return segment_iou(a, b)
+
+    def label_matcher(self, a, b):
+        a_label = self._context._get_any_label_name(a, a.label)
+        b_label = self._context._get_any_label_name(b, b.label)
+        return a_label == b_label
+
+    @classmethod
+    def _make_memoizing_distance(cls, distance_function: Callable[[Any, Any], float]):
+        distances = {}
+        notfound = object()
+
+        def memoizing_distance(a, b):
+            if isinstance(a, int) and isinstance(b, int):
+                key = (a, b)
+            else:
+                key = (id(a), id(b))
+
+            dist = distances.get(key, notfound)
+
+            if dist is notfound:
+                dist = distance_function(a, b)
+                distances[key] = dist
+
+            return dist
+
+        return memoizing_distance, distances
+
+    @staticmethod
+    def _get_ann_type(t, item: dm.Annotation) -> Sequence[dm.Annotation]:
+        return [a for a in item if a.type == t and not a.attributes.get("outside", False)]
+
+    def _match_segments(
+        self,
+        t,
+        item_a,
+        item_b,
+        *,
+        distance: Callable = distance,
+        label_matcher: Callable = None,
+        a_objs: Optional[Sequence[dm.Annotation]] = None,
+        b_objs: Optional[Sequence[dm.Annotation]] = None,
+        dist_thresh: Optional[float] = None,
+    ):
+        if label_matcher is None:
+            label_matcher = self.label_matcher
+        if dist_thresh is None:
+            dist_thresh = self.pairwise_dist
+        if a_objs is None:
+            a_objs = self._get_ann_type(t, item_a)
+        if b_objs is None:
+            b_objs = self._get_ann_type(t, item_b)
+
+        if self.return_distances:
+            distance, distances = self._make_memoizing_distance(distance)
+
+        if not a_objs and not b_objs:
+            distances = {}
+            returned_values = [], [], [], []
+        else:
+            extra_args = {}
+            if label_matcher:
+                extra_args["label_matcher"] = label_matcher
+
+            returned_values = match_segments(
+                a_objs,
+                b_objs,
+                distance=distance,
+                dist_thresh=dist_thresh if dist_thresh is not None else self.iou_threshold,
+                **extra_args,
+            )
+
+        if self.return_distances:
+            returned_values = returned_values + (distances,)
+
+        return returned_values
+
+    def match_annotations_two_sources(self, a, b):
```

Comment:
Please add output type info, it's not clear how this method should be defined in subclasses.